### PR TITLE
Implement taproot specific script helpers

### PR DESCRIPTION
**Pull request by jrawsthorne**
This change implements the script parts of https://github.com/rust-bitcoin/rust-bitcoin/issues/503. Will rebase with changes from https://github.com/rust-bitcoin/rust-bitcoin/pull/563 once it's merged

### CODE REVIEW

**Comment by sgeisler** [comment]
There is also a dependency on  #585 imo.

File: src/blockdata/script.rs [hunk 1]
```diff
@@ -271,6 +272,11 @@ impl Script {
         Script::new_witness_program(::bech32::u5::try_from_u8(0).unwrap(), &script_hash.to_vec())
     }
 
+    /// Generates P2TR-type of scriptPubkey
+    pub fn new_v1_tr(public_key: &schnorr::PublicKey) -> Script {
+        Script::new_witness_program(::bech32::u5::try_from_u8(1).unwrap(), &public_key.serialize().to_vec())
```
Comment (dr-orlovsky) [inline src/blockdata/script.rs:277]: Not sure why you are calling allocating `to_vec()` function just to take the reference after.

Also will need to get rid out of `u5`-caused `unwrap`, but that requires #563 merge as you said, with its `WitnessVersion` type

File: src/blockdata/script.rs [hunk 2]
```diff
@@ -395,6 +401,14 @@ impl Script {
             self.0[1] == opcodes::all::OP_PUSHBYTES_20.into_u8()
     }
 
+    /// Checks whether a script pubkey is a p2tr output
+    #[inline]
+    pub fn is_v1_p2tr(&self) -> bool {
+        self.0.len() == 34 &&
+        self.0[0] == opcodes::all::OP_PUSHNUM_1.into_u8() &&
+        self.0[1] == opcodes::all::OP_PUSHBYTES_32.into_u8()
```
Comment (dr-orlovsky) [inline src/blockdata/script.rs:409]: Not related to the PR, but just a thought that we probably need to refactor all these opcode-based version guessing via WitnessVersion API

File: src/blockdata/script.rs (hunk 1 above)
Comment (jrawsthorne) [reply src/blockdata/script.rs]: You're right. I've removed that and added a commit to do the same for new_v0_wpkh() and new_v0_wsh()

**Comment by dr-orlovsky** [comment]
Am I right that this PR was superseded by #585?

**Comment by sanket1729** [comment]
This is now done as a part of #696.

**Pull request closed without merging**